sdk/lib/rtl/heap.c

```diff
@@ -664,16 +664,21 @@ RtlpFindAndCommitPages(PHEAP Heap,
             /* Calculate first and last entries */
             FirstEntry = (PHEAP_ENTRY)Address;
 
-            /* Go through the entries to find the last one */
-            if (PreviousUcr)
-                LastEntry = (PHEAP_ENTRY)((ULONG_PTR)PreviousUcr->Address + PreviousUcr->Size);
-            else
-                LastEntry = &Segment->Entry;
-
-            while (!(LastEntry->Flags & HEAP_ENTRY_LAST_ENTRY))
+            LastEntry = Segment->LastEntryInSegment;
+            if (!(LastEntry->Flags & HEAP_ENTRY_LAST_ENTRY) ||
+                LastEntry + LastEntry->Size != FirstEntry)
             {
-                ASSERT(LastEntry->Size != 0);
-                LastEntry += LastEntry->Size;
+                /* Go through the entries to find the last one */
```

Comment:
It can still be used after any of the cases that sets `Segment->LastEntryInSegment = Segment->FirstEntry;`. There aren't many of those.
It would be possible to avoid doing this here by running the loop in those cases, but that's silly as it would be front-loading work that may not actually end up being necessary.